### ISSUE TITLE
16 bit data transfer.

### ISSUE THREAD

**Issue by peterjriggs** [david-res/ILI948x_t41_p#1]
Sorry if this is not the correct way to contact you over a matter like this.

I'm looking at this as a starting point for a project I have just started.
What needs to be altered for 16 bit data rather than 8 bit?
My programming is a bit rusty and I'm struggling to work through it, so hints would be appreciated.

**Comment by david-res** [comment]
Hi @peterjriggs 
There is some work to get this done, but not a huge effort.

First of all, you need to set the upper 8 FlexIO pins to outputs:
https://github.com/david-res/ILI948x_t41_p/blob/65a3fb57ab6b2a4f0944884b2b1d28727d0e138d/ILI948x_t41_p.cpp#L398

Then you need to set them to high strength using `*(portControlRegister(PIN_NUMBER) = 0xFF;` and them enable them as FlexIO pins using `pFlex->setIOPinToFlexMode(PIN_NUMBER);`

After that, you need to set the FlexIO registers to 16 bit bus width, so any place you see `FLEXIO_SHIFTCFG_PWIDTH(7);` needs to be changed to `FLEXIO_SHIFTCFG_PWIDTH(15);`

You need to then alter the functions that push the data out, such as `SglBeatWR_nPrm_16`
It currently pushes 8 bits at a time like so:

`{
      for(uint32_t i=0; i<length-1U; i++)
        {
          buf = *value++;
            while(0 == (p->SHIFTSTAT & (1U << 0)))
            {
            }
            p->SHIFTBUF[0] = buf >> 8;

            while(0 == (p->SHIFTSTAT & (1U << 0)))
            {
            }
            p->SHIFTBUF[0] = buf & 0xFF;
        }
        buf = *value++;
        //Write the last byte 
        while(0 == (p->SHIFTSTAT & (1U << 0)))
            {
            }
        p->SHIFTBUF[0] = buf >> 8;

        while(0 == (p->SHIFTSTAT & (1U << 0)))
        {
        }
        p->TIMSTAT |= (1U << 0);

        p->SHIFTBUF[0] = buf & 0xFF;

        /*Wait for transfer to be completed */
        while(0 == (p->TIMSTAT |= (1U << 0)))
        {
        }
}`


But would need to be changed to something in the lines of:

`{
      for(uint32_t i=0; i<length-1U; i++)
        {
            while(0 == (p->SHIFTSTAT & (1U << 0)))
            {
            }
            p->SHIFTBUF[0] = *value++;
        }
        //Write the last byte 
        while(0 == (p->SHIFTSTAT & (1U << 0)))
            {
            }
        p->TIMSTAT |= (1U << 0);

        p->SHIFTBUF[0] = *value++;

        /*Wait for transfer to be completed */
        while(0 == (p->TIMSTAT |= (1U << 0)))
        {
        }`
        
When writing commands `SglBeatWR_nPrm_8` will also need to be modified in a similar way, and you might need to bit shift the command and data by 8 bits

Is it possible for you to run on 8 bits for now, and wait a couple of weeks for me to add support for 16 bits, or would you rather have a go at it yourself?

**Comment by peterjriggs** [comment]
david-res,
Thank you so much for that.
I'm more used to devices where parallel output is somewhat more direct. I had worked out the configuration side but when that was not enough I was a bit lost.
I have tested your suggestion and am now getting 16 bit output.

My immediate application is to drive a 16 bit DAC so the 8 bit command writes are not required.
I can now see if I can control the output timing sufficiently for my needs.
Thanks again ...
